ignite/engine.py

```diff
@@ -93,44 +99,34 @@ def decorator(f):
             return f
         return decorator
 
-    def _fire_event(self, event_name, *event_args):
+    def _fire_event(self, event_name, state, *event_args):
         if event_name in self._event_handlers.keys():
             self._logger.debug("firing handlers for event %s ", event_name)
             for func, args, kwargs in self._event_handlers[event_name]:
-                func(self, *(event_args + args), **kwargs)
-
-    def terminate(self):
```

Comment:
hmm this causes some issues. As I feel `terminate` should really be a method on the engine. I see the work around below of setting a `terminate` attribute on the state, but it really feels more natural to have it on the `Engine` paired with `run`

Reply:
This can be kept here if we keep the state private to the engine and expose it via a `get_state()` function, then the handlers still get passed the engine and can access the state (and the method `get_state()` is more obvious than just having it as an attribute)

Reply:
My thinking is that state is a property of the specific "run", not of the trainer itself. Similarly, you are terminating the current run rather than the engine/trainer itself. That being said, i'm definitely ok with reverting the terminate changes. I don't feel strongly one way or the other there.

I would like to keep it as the state being passed to handlers though as I think that makes the most sense.

What do you think? Should I just revert the terminate change?

Reply:
By just reverting the terminate change, how would a user call `terminate()` on the engine from within a handler?

Reply:
The way I have it being done in the tests is the handlers close over the trainer. So they can just call trainer.terminate() without needed it passed in. I recognize this is less than ideal though.

Here are the options I see:

1. Keep it the way it is in this PR. Handlers can toggle `state.terminate` in order to terminate the run.
2. add a `State.terminate()` method that essentially does what `Trainer#terminate()` did before
3. revert to previous `Trainer#terminate()` approach and require that handlers close over the trainer if they need to call this
4. revert to previous `Trainer#terminate()` approach and pass two arguments to each handler: trainer and state.
5. revert to previous `Trainer#terminate()` approach and set state as an attribute on Trainer. Handlers get passed the trainer.

Personally I like option 1. But if we feel like `terminate` belongs on the `Engine` itself, then I would vote for option 4. The only option I really would like to avoid is option 5. Since state is ephemeral and specific to that run, I don't think it should be an attribute of the trainer. Also, most handlers are gonna be interacting with the state object, so it'd be nice to avoid having to go through the trainer to do this.

What do you think?

Reply:
Personally I don't think terminate belongs on the state as it logically pairs with `run`. Additionally doing `state.terminate = True` is not obvious that it should effect change in the engine. 

`terminate()` will probably be called quite a bit for various early stopping handlers, so the usage pattern has to be 1) easy and 2) intuitive. The only problem with closing over the trainer is I'm not sure how intuitive this is. 

I also agree that passing 2 args to every handler when mostly you will use one or the other is not ideal, and neither is having to do `state = trainer.get_state()` at the start of every handler. 

I'm ok with 3/4/5 but neither stand out as clear choices. Also 2 could work but calling/reading `state.terminate()` still isn't super clear to me. 

I'll try and think more about this tonight - but I don't think it's a huge problem, either of them will work we just need to make it clear in examples

Reply:
All good points. I think you make a strong argument not to use option 1 or 2. I'm hesitant to use option 3 as I agree that it will likely be very unintuitive for users.

I actually think option 4 works quite well though. Conceptually it makes sense to me that in a handler i am passed both the engine itself and the current state of the run. Its also the easiest from user education standpoint, IMO.

Reply:
Let's go with 4 then, I think it's a bit more verbose but will be very clear from the API what's going on.